scene_synchronizer.cpp

```diff
@@ -3239,196 +3210,259 @@ void ClientSynchronizer::process_controllers_recovery(real_t p_delta) {
+					player_controller_node_data->dependency_nodes.find(rew_node_data) != -1) {
+				// Controller node.
+				r_recover_controller = true;
+			} else {
+				r_nodes_to_recover.push_back(rew_node_data);
+			}
+		}
 
-		const int remaining_inputs = player_controller->notify_input_checked(checkable_input_id);
 #ifdef DEBUG_ENABLED
-		// Unreachable because the SceneSynchronizer and the PlayerController
-		// have the same stored data at this point.
-		CRASH_COND(client_snapshots.size() != size_t(remaining_inputs));
-#endif
+		if (different) {
+			// Emit the de-sync detected signal.
 
-#ifdef DEBUG_ENABLED
-		// Used to double check all the instants have been processed.
-		bool has_next = false;
-#endif
-		for (int i = 0; i < remaining_inputs; i += 1) {
-			scene_synchronizer->change_events_begin(NetEventFlag::SYNC_RECOVER | NetEventFlag::SYNC_REWIND);
+			static const Vector<NetUtility::Var> const_empty_vector;
+			const Vector<NetUtility::Var> &server_node_vars = uint32_t(server_snapshots.front().node_vars.size()) <= net_node_id ? const_empty_vector : server_snapshots.front().node_vars[net_node_id];
+			const Vector<NetUtility::Var> &client_node_vars = uint32_t(client_snapshots.front().node_vars.size()) <= net_node_id ? const_empty_vector : client_snapshots.front().node_vars[net_node_id];
 
-			// Step 1 -- Process the scene nodes.
-			for (uint32_t r = 0; r < nodes_to_recover.size(); r += 1) {
-				if (nodes_to_recover[r]->sync_enabled == false) {
-					// This node is not sync.
-					continue;
+			const int count = MAX(server_node_vars.size(), client_node_vars.size());
+
+			variable_names.resize(count);
+			server_values.resize(count);
+			client_values.resize(count);
+
+			for (int g = 0; g < count; ++g) {
+				if (g < server_node_vars.size()) {
+					variable_names.ptrw()[g] = server_node_vars[g].name;
+					server_values.ptrw()[g] = server_node_vars[g].value;
+				} else {
+					server_values.ptrw()[g] = Variant();
 				}
-				nodes_to_recover[r]->process(p_delta);
-#ifdef DEBUG_ENABLED
-				if (nodes_to_recover[r]->functions.size()) {
-					SceneSynchronizerDebugger::singleton()->debug_print(scene_synchronizer, "Rewind, processed node: " + nodes_to_recover[r]->node->get_path());
+
+				if (g < client_node_vars.size()) {
+					variable_names.ptrw()[g] = client_node_vars[g].name;
+					client_values.ptrw()[g] = client_node_vars[g].value;
+				} else {
+					client_values.ptrw()[g] = Variant();
 				}
-#endif
 			}
 
-			// Step 2 -- Process the controller.
-			if (recover_controller && player_controller_node_data->sync_enabled) {
-#ifdef DEBUG_ENABLED
-				has_next =
-#endif
-						controller->process_instant(i, p_delta);
-				SceneSynchronizerDebugger::singleton()->debug_print(scene_synchronizer, "Rewind, processed controller: " + controller->get_path());
+			if (rew_node_data->node) {
+				scene_synchronizer->emit_signal("desync_detected", p_input_id, rew_node_data->node, variable_names, client_values, server_values);
```

Comment:
Does this desync signal expose the input buffer as well the id? I think we'll want the buffer data, perhaps a particular input type is resulting in the desyncs, we'll need an API to get the input buffer from the id, if one doesn't already exist.